implementations/rust/ockam/ockam_command/src/enroll/command.rs

```diff
@@ -0,0 +1,371 @@
+            .to_string()
+            .color(OckamColor::PrimaryResource.color())
+    ))?;
+    Ok(())
+}
+
+async fn retrieve_user_project(
+    ctx: &Context,
+    opts: &CommandGlobalOpts,
+    node_name: &str,
+) -> Result<IdentityIdentifier> {
+    let space = default_space(ctx, opts, node_name)
+        .await
+        .wrap_err("Unable to retrieve and set a space as default")?;
+    info!("Retrieved the user default space {:?}", space);
+
+    let project = default_project(ctx, opts, node_name, &space)
+        .await
+        .wrap_err(format!(
+            "Unable to retrieve and set a project as default with space {}",
+            space
+                .name
+                .to_string()
+                .color(OckamColor::PrimaryResource.color())
+        ))?;
+    info!("Retrieved the user default project {:?}", project);
+
+    let identifier = update_enrolled_identity(opts, node_name)
+        .await
+        .wrap_err(format!(
+            "Unable to set the local identity as enrolled with project {}",
+            project
+                .name
+                .to_string()
+                .color(OckamColor::PrimaryResource.color())
+        ))?;
+    info!("Enrolled a user with the IdentityIdentifier {}", identifier);
+
+    Ok(identifier)
+}
+
+/// Enroll a user with a token, using a specific node to contact the controller
+pub async fn enroll_with_node(
+    ctx: &Context,
+    opts: &CommandGlobalOpts,
+    route: &MultiAddr,
+    node_name: &str,
+    token: Auth0Token,
+) -> miette::Result<()> {
+    let mut rpc = RpcBuilder::new(ctx, opts, node_name).build();
+    rpc.request(api::enroll::auth0(route, token)).await?;
+    let (res, dec) = rpc.check_response()?;
+    if res.status() == Some(Status::Ok) {
+        info!("Enrolled successfully");
+        Ok(())
+    } else if res.status() == Some(Status::BadRequest) {
+        info!("Already enrolled");
+        Ok(())
+    } else {
+        Err(miette!("{}", rpc.parse_err_msg(res, dec)))
+    }
+}
+
+/// Use an embedded node to enroll a user with a token
+pub async fn enroll(
+    ctx: &Context,
+    opts: &CommandGlobalOpts,
+    token: Auth0Token,
+) -> miette::Result<()> {
+    let node_name = start_embedded_node(ctx, opts, None).await?;
+    enroll_with_node(ctx, opts, &CloudOpts::route(), node_name.as_str(), token).await?;
+    retrieve_user_project(ctx, opts, &node_name)
+        .await
+        .map(|_| ())
+        .into_diagnostic()
+}
+
+async fn default_space<'a>(
+    ctx: &Context,
+    opts: &CommandGlobalOpts,
+    node_name: &str,
+) -> Result<Space<'a>> {
+    // Get available spaces for node's identity
+    opts.terminal
+        .write_line(&fmt_log!("Getting available spaces in your account..."))?;
+
+    let mut rpc = RpcBuilder::new(ctx, opts, node_name).build();
+    let is_finished = Mutex::new(false);
+    let send_req = async {
+        rpc.request(api::space::list(&CloudOpts::route())).await?;
+        *is_finished.lock().await = true;
```

Comment:
Would a `std::sync::atomic::AtomicBool` also work, and would it have any advantages?

Reply:
Good question. I think this would more to the point, rather than having a `Mutex` on a `bool`. (I just moved that code around in that PR though, I did not touch it)